src/main/kotlin/org/rust/ide/ssr/RsMatchingVisitor.kt

```diff
@@ -0,0 +1,258 @@
+        myMatchingVisitor.result =
+            matchOuterAttrList(o.outerAttrList, field.outerAttrList) &&
+                match(o.vis, field.vis) &&
+                matchIdentifier(o.identifier, field.identifier) &&
+                match(o.typeReference, field.typeReference)
+    }
+
+    override fun visitTupleFields(o: RsTupleFields) {
+        val fields = getElement<RsTupleFields>() ?: return
+        myMatchingVisitor.result = myMatchingVisitor.matchSequentially(o.tupleFieldDeclList, fields.tupleFieldDeclList)
+    }
+
+    override fun visitTupleFieldDecl(o: RsTupleFieldDecl) {
+        val field = getElement<RsTupleFieldDecl>() ?: return
+        myMatchingVisitor.result =
+            matchOuterAttrList(o.outerAttrList, field.outerAttrList) &&
+                match(o.vis, field.vis) &&
+                match(o.typeReference, field.typeReference)
+    }
+
+    override fun visitTypeReference(o: RsTypeReference) {
+        val typeReference = getElement<RsTypeReference>() ?: return
+        // TODO: implement individual type references
+        myMatchingVisitor.result = matchTextOrVariable(o, typeReference)
+    }
+
+    override fun visitRefLikeType(o: RsRefLikeType) {
+        val refType = getElement<RsRefLikeType>() ?: return
+        myMatchingVisitor.result = (o.mut == null) == (refType.mut == null) &&
+                (o.const == null) == (refType.const == null) &&
+                (o.mul == null) == (refType.mul == null) &&
+                match(o.typeReference, refType.typeReference) &&
+                match(o.lifetime, refType.lifetime)
+    }
+
+    override fun visitLifetime(o: RsLifetime) {
+        val lifetime = getElement<RsLifetime>() ?: return
+        myMatchingVisitor.result = matchTextOrVariable(o, lifetime)
+    }
+
+    override fun visitVis(o: RsVis) {
+        val vis = getElement<RsVis>() ?: return
+        myMatchingVisitor.result = matchTextOrVariable(o, vis)
+    }
+
+    override fun visitInnerAttr(o: RsInnerAttr) {
+        val attr = getElement<RsInnerAttr>() ?: return
+        myMatchingVisitor.result = match(o.metaItem, attr.metaItem)
+    }
+
+    override fun visitOuterAttr(o: RsOuterAttr) {
+        val attr = getElement<RsOuterAttr>() ?: return
+        myMatchingVisitor.result = match(o.metaItem, attr.metaItem)
+    }
+
+    override fun visitMetaItem(o: RsMetaItem) {
+        val metaItem = getElement<RsMetaItem>() ?: return
+        myMatchingVisitor.result = match(o.compactTT, metaItem.compactTT) &&
+            match(o.litExpr, metaItem.litExpr) &&
+            match(o.metaItemArgs, metaItem.metaItemArgs) &&
+            match(o.path, metaItem.path)
+    }
+
+    override fun visitMetaItemArgs(o: RsMetaItemArgs) {
+        val metaItemArgs = getElement<RsMetaItemArgs>() ?: return
+        myMatchingVisitor.result = myMatchingVisitor.matchInAnyOrder(o.metaItemList, metaItemArgs.metaItemList)
+    }
+
+    override fun visitPathType(o: RsPathType) {
```

Comment:
Why treat types specially here? Probably it will be unintuitive that paths in types and other positions are matched differently. E.g. consider this example with `RsPathType` and `RsPathExpr`:
```rust
struct Struct(i32);

let x: Struct = Struct(1);
       ~~~~~~   ~~~~~~
       type     pathExpr
```

Reply:
I moved the custom logic to `visitPath`, it can indeed apply to all paths.